### PR TITLE
faster index page, index by prefix, etc

### PR DESCRIPTION
**Pull request by crawshaw**
Please see commit descriptions for details.

### CODE REVIEW

**Comment by crawshaw** [comment]
CC @adg @campoy

**Comment by crawshaw** [comment]
Any prefix index is a subset of http://godoc.org/-/index (in fact that's how the code generates it), so as long as we can support that we don't have to worry about the prefixes.

**Comment by garyburd** [comment]
Does this PR make the index page faster in practice? There will be frequent cache refreshes if db.Put is called more frequently than db.Index.

This is the first code in the project that assumes that all writes to the database are from a single process. 

The new Redis *SCAN commands can be used to implement prefix search. 

The /-/index page is huge.  Caching the index in memory makes it possible to implement alternatives to the  /-/index such as prefix match in this PR and fuzzy matching.

**Comment by crawshaw** [comment]
The Etag helps a lot.

It does assume the caching server is doing database Puts. But other than that, the worse that can happen is indexes get a little out of date.

I haven't tested the redis timing or the GC pressure saved because I don't have a copy of the full database. (I'll ping someone on Monday for access.) My suspicion is it will help because right now each hit to the index page generates a data structure with ~150k pointers.

Mostly, keeping the index in memory just makes the prefix searching really easy. At 70k packages we are at ~2mb of strings, so this should be good for the foreseeable future.

**Comment by garyburd** [comment]
When I ran the service, db.Put was called more frequently than db.Index. I suspect that this is still true. 

The Etag will definitely helpful.

**Comment by crawshaw** [comment]
If that's still true, then the cache doesn't help much. But it's pretty hard to guess at these things -- making the index faster may mean it gets used more, and the prefix index feature means it will get used more. (For example, right now if you hit http://rsc.io you get a 404, after this change that would turn into an index.)

Unless there's some fundamental objection to this strategy, I'd like to try it.

**Comment by garyburd** [comment]
Because the time to transmit the /-/index page to the client is large compared to the time it takes to fetch the data from Redis, I doubt that the cache will increase the use of this page. 

I suggest instrumenting the code to determine the cache hit rate on the production server.

**Comment by crawshaw** [comment]
Depending on what comes up, I suspect I won't get to instrumentation for a couple of weeks. But just to state my assumptions: I agree that asking redis for a couple of megabytes should be trivial. It's the step where we turn it into 150k strings and throw them away I suspect might cost a bit.

But I didn't really do this with speeding up /-/index in mind. I did it with speeding up /github.com and the like in mind, as they all use the same cached copy. (Also, it was really easy.)

**Comment by garyburd** [comment]
Prefix queries can be implemented in Redis using lexical range queries on a sorted set. If prefix queries are  implemented, then code specific specific to the "all", "project" and "subrepo" indexes can be removed. I did not use lexical range queries because  Redis did not have the feature at the time I wrote the code.

In a previous comment I said that the SCAN commands can be used to to implement prefix queries. That's true, but it's less efficient and not any easier than implementing the feature with ZRANGEBYLEX.

All of the queries implemented in Redis can be moved to the Go application. There are benefits to moving the features to the Go application. It will be easier for Go developers to contribute features.  Features are not constrained by what's possible or easy with Redis.

This PR implements in Go code a more general form of the project and subrepo queries implemented in Redis. I don't like the approach of implementing similar functionality in different ways.

I suggest accepting this PR and opening an issue on how queries should be implemented. Should we implement queries in Redis, or should we move everything to Go?

I think performance concerns about /-/index, /github.com and other unwieldy pages is secondary to the concern about how the code is structured.  Implementing these pages using an in memory cache will not improve the performance enough to make the pages more useful. The in memory cache is not required to implement etags. The in memory cache will reduce GC pauses, but that's less of an issue with Go 1.5.

File: database/database.go [hunk 1]
```diff
@@ -74,6 +82,17 @@ var (
 	redisLog         = flag.Bool("db-log", false, "Log database commands")
 )
 
+func pkgsHash(pkgs []Package) string {
+	nl := []byte{'\n'}
+	h := fnv.New64a()
+	for _, p := range pkgs {
+		io.WriteString(h, p.Path)
```
Comment (campoy) [inline database/database.go]: why not:

fmt.Fprintf(h, "%s%s\n", p.Path, p.Synopsis)

File: database/database.go [hunk 2]
```diff
@@ -571,8 +596,45 @@ func (db *Database) GoSubrepoIndex() ([]Package, error) {
 	return db.getPackages("index:project:subrepo", false)
 }
 
-func (db *Database) Index() ([]Package, error) {
-	return db.getPackages("index:all:", false)
+func (db *Database) indexLoad() (index []Package, indexHash string, err error) {
+	db.indexMu.Lock()
```
Comment (campoy) [inline database/database.go]: would it be worth to have a RWMutex?

**Comment by crawshaw** [comment]
@garyburd I agree that it should be either all in redis or all in Go, and I'd like to put a vote in for doing it in Go. I don't think redis buys us anything, and makes some things more complex. (I have a simple pagerank implementation for package ranking that's 100 lines of Go. It would be more in redis, if it's even possible.)

I'll circle back on this in a couple weeks when I have more time.

**Comment by adg** [comment]
I agree it would be nice to drop redis from the equation.

**Comment by shurcooL** [comment]
I also agree and I'm in favor of doing it in Go (fwiw).

**Comment by adg** [comment]
@crawshaw We've switched to Gerrit for the gddo repo now. Should I just close this?

**Pull request closed without merging**